internal/tools/cloudflare/dns_test.go

```diff
@@ -0,0 +1,506 @@
+			},
+			setupZoneID: func(zoneName string) (zoneID string, err error) {
+				return "", errors.New("failed to get zone ID")
+			},
+			setupRecordID: func(zoneID, customerDNSName string, logger logrus.FieldLogger) (recordID string, err error) {
+				return "", nil
+			},
+			setupDeleteDNSRecord: func(ctx context.Context, zoneID, recordID string) error {
+				return nil
+			},
+			setupDNSRecord: func(ctx context.Context, zoneID string, rr cf.DNSRecord) ([]cf.DNSRecord, error) {
+				return []cf.DNSRecord{}, nil
+			},
+			expected: errors.New("failed to get zone ID"),
+		},
+		{
+			description:     "failure to get record ID",
+			customerDNSName: "customer.cloud.mattermost.com",
+			zoneNameList:    []string{"cloud.mattermost.com"},
+			zoneID:          "RANDOMDIDFROMCLOUDFLARE",
+			setupName: func(zoneNameList []string, customerDNSName string) (zoneName string, found bool) {
+				return "cloud.mattermost.com", true
+			},
+			setupZoneID: func(zoneName string) (zoneID string, err error) {
+				return "RANDOMDIDFROMCLOUDFLARE", nil
+			},
+			setupRecordID: func(zoneID, customerDNSName string, logger logrus.FieldLogger) (recordID string, err error) {
+				return "", errors.New("failed to get record ID")
+			},
+			setupDeleteDNSRecord: func(ctx context.Context, zoneID, recordID string) error {
+				return nil
+			},
+			setupDNSRecord: func(ctx context.Context, zoneID string, rr cf.DNSRecord) ([]cf.DNSRecord, error) {
+				return []cf.DNSRecord{}, nil
+			},
+			expected: errors.New("failed to get record ID"),
+		},
+		{
+			description:     "failure to delete DNS record",
+			customerDNSName: "customer.cloud.mattermost.com",
+			zoneNameList:    []string{"cloud.mattermost.com"},
+			zoneID:          "RANDOMDIDFROMCLOUDFLARE",
+			setupName: func(zoneNameList []string, customerDNSName string) (zoneName string, found bool) {
+				return "cloud.mattermost.com", true
+			},
+			setupZoneID: func(zoneName string) (zoneID string, err error) {
+				return "RANDOMDIDFROMCLOUDFLARE", nil
+			},
+			setupRecordID: func(zoneID, customerDNSName string, logger logrus.FieldLogger) (recordID string, err error) {
+				return "RANDOMRECORDIDFROMCLOUDFLARE", nil
+			},
+			setupDeleteDNSRecord: func(ctx context.Context, zoneID, recordID string) error {
+				return errors.New("failed to delete DNS record")
+			},
+			setupDNSRecord: func(ctx context.Context, zoneID string, rr cf.DNSRecord) ([]cf.DNSRecord, error) {
+				return []cf.DNSRecord{}, nil
+			},
+			expected: errors.New("failed to delete DNS record"),
+		},
+	}
+
+	for _, s := range samples {
+		t.Run(s.description, func(t *testing.T) {
+			mockCF.mockGetZoneName = s.setupName
+			mockCF.mockGetZoneID = s.setupZoneID
+			mockCF.mockGetRecordID = s.setupRecordID
+			mockCF.mockDeleteDNSRecord = s.setupDeleteDNSRecord
+			mockCF.mockDNSRecords = s.setupDNSRecord
+			client := NewClientWithToken(mockCF)
+			err := client.DeleteDNSRecord(s.customerDNSName, s.zoneNameList, logger)
+			if err != nil {
```

Comment:
What if we expect an error but got nil? The test would pass in such a case but it should not.
```suggestion
			if s.expected != nil {
```